### ISSUE TITLE
SUNDIALS interface.

### ISSUE THREAD

**Issue by luca-heltai** [dealii/dealii#11435]
Issue to keep track of SUNDIALS update related issues.

- [ ] update sundials in spack (fyi @jppelteret)
- [ ] make sure all interfaces return internal objects as the actual types, i.e., IDAMem, ARKOdeMem, etc.
- [ ] move to a seperate header the interfaces for NVector, SUNMatrix, etc.
- [ ] introduce interfaces to SUNMatrix, SUNLinsolver, etc. 
- [ ] adapt interfaces to existing deal.II classes (i.e., LinearOperator)

**Comment by luca-heltai** [comment]
#11395

**Comment by jppelteret** [comment]
Thanks for the heads up @luca-heltai. Thankfully, it looks like the spack Sundials package is [already up-to-date](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/sundials/package.py#L25). Should be an [easy change for us](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dealii/package.py#L183):
```
depends_on('sundials@:3~pthread', when='@9.0:9.2+sundials')
depends_on('sundials@5:~pthread', when='@9.3:+sundials') # Or sundials@5.X ; not sure if the ~pthread restriction remains
```

**Comment by sebproell** [comment]
>  make sure all interfaces return internal objects as the actual types, i.e., IDAMem, ARKOdeMem, etc.

I just remembered: in the new versions these types are not know to us as they are hidden in implementation files that are not part of the installed headers.

One more thing I would like to do: currently, I can only set up our `ARKode` object and then call the `solve_ode` function which performs a complete time loop. I would also like to only evaluate a single step inside my own time loop. This can be achieved easily by extracting the single step solve into a public method.

**Comment by luca-heltai** [comment]
@jppelteret, I'm testing 

~~
epends_on('sundials@:3~pthread', when='@9.0:9.2+sundials')
depends_on('sundials@5:~pthread', when='@9.3:+sundials') # Or sundials@5.X ; not sure if the ~pthread restriction remains
~~~

do you remember why we needed `sundials~pthread`? I tried leaving out the `~pthread`, and so far everything seems to work fine...

**Comment by luca-heltai** [comment]
https://github.com/spack/spack/pull/20621

**Comment by jppelteret** [comment]
> do you remember why we needed sundials~pthread? I tried leaving out the ~pthread, and so far everything seems to work fine..

No, I'm afraid have no idea. Could there be a conflict with our TBB? I'd say that if things seem to be working without the restriction then we should remove it. We can always add it back again if there are some troubles.

**Comment by sebproell** [comment]
I refactored the current implementation of the linear solver a bit in #11724.

@luca-heltai Do you have suggestions how dealii's interface to SUNLinearSolver should look in the long run? Currently it only supprts iterative solvers and wraps a user's `LinearSolveFunction`.  The fact that this wrapper is even created is hidden in the implementation. Still, users need to write a small lambda to wrap a dealii solver. Do you think we should introduce an easier way to directly use a solver object? Keep in mind that SUNDIALS passes a dynamically calculated solver tolerance when requesting a solve.
